lib/scan_result.dart

```diff
@@ -14,49 +14,66 @@ abstract class _ScanResultMetadata {
 
 /// A scan result emitted by the scanning operation, containing [Peripheral] and [AdvertisementData].
 class ScanResult {
-  Peripheral peripheral;
+  final Peripheral peripheral;
 
   /// Signal strength of the peripheral in dBm.
-  int rssi;
+  final int rssi;
 
   /// An indicator whether the peripheral is connectable (iOS only).
-  bool isConnectable;
+  late final bool isConnectable;
 
   /// A list of UUIDs found in the overflow area of the advertisement data (iOS only).
-  List<String> overflowServiceUuids;
+  late final List<String> overflowServiceUuids;
 
   /// A packet of data advertised by the peripheral.
-  AdvertisementData advertisementData;
-
-  ScanResult.fromJson(Map<String, dynamic> json, ManagerForPeripheral manager)
-      : peripheral = Peripheral.fromJson(json, manager),
-        rssi = json[_ScanResultMetadata.rssi],
-        isConnectable = json[_ScanResultMetadata.isConnectable],
-        overflowServiceUuids = json[_ScanResultMetadata.overflowServiceUuids],
-        advertisementData = AdvertisementData._fromJson(json);
+  final AdvertisementData advertisementData;
+  
+  ScanResult._(
+    this.peripheral, 
+    this.rssi,
+    this.advertisementData,
+    {bool? isConnectable, 
+    List<String>? overflowServiceUuids, 
+  }) : isConnectable = isConnectable ?? false,
+       overflowServiceUuids = overflowServiceUuids ?? <String>[];
+
+
+  factory ScanResult.fromJson(
+    Map<String, dynamic?> json, 
+    ManagerForPeripheral manager
+  ) {
+    assert(json[_ScanResultMetadata.rssi] is int);
```

Comment:
motivations:
-easier to mock a dummy ScanResult object, you just need a @visibleForTesting on a static method that calls this private constructor.
-more importantly, type checker catches _TyepeError-s at the right place
-if something is null out of the dict object it can be concisely handled there in the constructor
-if you want to avoid late qualifiers, the fields must be initialised in the initialiser list, the constructor body -- here factory body -- is too late, so I wanted to avoid to having to us `late` as semantically it's not really late is it?

Reply:
But the original constructor didn't have a body and was public? The only thing changed is (apart from the assert) `overflowServiceUuids ?? <String>[]`.

Doesn't really matter, as this should be changed to abstract class in the future with the implementation moved to src/. Marking as resolved.